### PR TITLE
Fix moment commutation detection for group-commuting operations (#6659)

### LINKED ISSUE #6659: Issue when commuting moments
**Description of the issue**
Cirq gives the wrong result when determining whether moments commute if the individual operators within the moments do not commute, but the moment as a whole still commutes. For example, take a pair of single qubit Z gates and a two qubit RXX gate. While [Z, RXX] !=0 (for either single qubit Z), we do in fact have [Z * Z, RXX] = 0. However, when using cirq we have

```
import cirq
qubits = cirq.LineQubits.range(2)
moment_1 = cirq.Moment([cirq.Z(qubits[0], cirq.Z(qubits[1])])
moment_2 = cirq.XXPowGate(exponent=1 / 2)(*qubits)

print(cirq.commutes(moment_1, moment_2))
```
Gives `False`.

Looking through the source code this seems to be because the the `commutes()` function compares pairs of operators form each moment but doesn't consider the case when the moment as a whole commutes even when the individual operators do not.

**How to reproduce the issue**
See above

**Cirq version**
You can get the cirq version by printing `cirq.__version__`. From the command line:
1.5.0

### PR DESCRIPTION
**Pull request by chamodtharakaperera**
Fixes #6659

This PR fixes an issue where cirq.commutes() incorrectly returns False for moments that commute as a whole even when their individual operations don't commute. For example, given two Z gates and an RXX gate, [Z⊗Z, RXX] = 0 even though neither Z commutes with RXX individually.

The solution:
- First tries fast path of checking individual operation commutation
- If individual operations don't commute or result is indeterminate, falls back to checking if the full moments commute via unitary calculation
- Properly handles edge cases:
  - Disjoint qubit sets (trivially commute)
  - Non-unitary operations (returns NotImplemented)
  - Indeterminate cases (returns NotImplemented)

Added comprehensive test cases covering:
- Original Z⊗Z with RXX commutation case
- Disjoint qubit operations
- Mixed cases where some operations commute individually and others as a group
- Non-commuting cases to verify negative results
- Complex cases requiring unitary calculation

Test outputs demonstrate correct behavior for all cases including the original issue where Z⊗Z should commute with RXX.

### CODE REVIEW

File: cirq-core/cirq/circuits/moment.py [hunk 1]
```diff
@@ -659,37 +659,72 @@ def cleanup_key(key: Any) -> Any:
...
+                break
+            if not commutes:
+                all_commute = False
+                break
+
+        if all_commute is True:
+            return True
+
+        # If individual operations don't commute or we couldn't determine,
+        # try checking if the full moments commute using unitary matrices
```
Comment (daxfohl) [inline cirq-core/cirq/circuits/moment.py:714]: Seems like this could be part of `protocols.commutes` itself. If that was the case then we'd have full coverage of stuff that commutes or not, at least for unitary things, but we can still override `_commutes_` like here for perf reasons. @pavoljuhas wdyt?

Comment (pavoljuhas) [reply cirq-core/cirq/circuits/moment.py:714]: working on it today, update soon ...

**Comment by pavoljuhas** [comment]
The PR was failing some tests as they were initially on the main branch
```
git checkout a0b813812f41e5b2216c7794f0399451bdf44785
git checkout "HEAD^2" -- cirq-core/cirq/circuits/moment_test.py
check/pytest -n0 cirq-core/cirq/circuits/moment_test.py

...
FAILED cirq-core/cirq/circuits/moment_test.py::test_commutes - TypeError: Failed to determine whether or not cirq.Moment(
```
The [Operation._commutes_](https://github.com/quantumlib/Cirq/blob/c3522c6fbd49199515cebba8c28e411f7fcdd664/cirq-core/cirq/ops/raw_types.py#L692-L718) function is already doing some clever things like checking the measurement and control keys, as well, as making sure calculation of unitaries would not take too much memory.

I decided to reuse that code by converting Moment to CircuitOperation if needed and then falling back to the commutes evaluation for operations.  Here is my take on it,

@chamodtharakaperera and @daxfohl - please let me know if this looks OK

**Comment by daxfohl** [comment]
Should that logic from `Operation._commutes_` be moved into the commutes protocol itself? Usually when I see jumping through hoops to use logic from one class in another class, it means that the logic was in the wrong place to begin with. I'd say either move it up a layer (into the commutes protocol as a strategy) or down a layer into a shared library function. As is, I think it adds too many implicit dependencies and I'm also not sure there's not an infinite recursion in there somewhere.

**Comment by pavoljuhas** [comment]
> Should that logic from `Operation._commutes_` be moved into the commutes protocol itself? Usually when I see jumping through hoops to use logic from one class in another class, it means that the logic was in the wrong place to begin with. I'd say either move it up a layer (into the commutes protocol as a strategy) or down a layer into a shared library function. As is, I think it adds too many implicit dependencies and I'm also not sure there's not an infinite recursion in there somewhere.

@daxfohl - I feel it is better to keep the protocol implementation at its top level general and isolated from the details (and import dependencies) of Operation / Moment classes.  The second suggestion sounds very good, here is my take on it - 2fee5948f61d68bd84bf271fa54b2c9a41989484...a55b67516d6033028d360ca6c16c8bd1a822aa9e.
Can you PTAL?

File: cirq-core/cirq/ops/raw_types.py [hunk 2]
```diff
@@ -1112,3 +1078,78 @@ def _validate_qid_shape(val: Any, qubits: Sequence['cirq.Qid']) -> None:
...
+    ops1_control_keys = frozenset(k for op in ops1 for k in protocols.control_keys(op))
+    ops2_control_keys = frozenset(k for op in ops2 for k in protocols.control_keys(op))
+    if (
+        not ops1_keys.isdisjoint(ops2_keys)
+        or not ops1_control_keys.isdisjoint(ops2_keys)
+        or not ops2_control_keys.isdisjoint(ops1_keys)
+    ):
+        return False
+
+    ops1_qubits = frozenset(q for op in ops1 for q in op.qubits)
```
Comment (daxfohl) [inline cirq-core/cirq/ops/raw_types.py:1117]: Kind of odd the disparity between qubits gotten via public member methods and cbits gotten from protocols. Seems like it should be consistent. (Preferably use members for speed). (Ultimately I think all those measurement key protocols should be phased out. They're not really what protocols were designed for. IIRC back when I looked into doing so, their use within the commutes protocol was a blocker for that. So if it's possible to avoid them here that would be nice).

**Review by daxfohl — APPROVED** [review]
Looks good. I added a note, but may be out of scope here. But if it's a simple change I think it'd be worth it.

File: cirq-core/cirq/circuits/moment.py [hunk 3]
```diff
@@ -659,37 +659,26 @@ def cleanup_key(key: Any) -> Any:
...
-            if not other_qubits.intersection(set(op.qubits)):
-                continue
-
-            commutes = protocols.commutes(op, other, atol=atol, default=NotImplemented)
-
-            if not commutes or commutes is NotImplemented:
-                return commutes
-
-        return True
+        other_operations = other.operations if isinstance(other, Moment) else (other,)
```
Comment (daxfohl) [inline cirq-core/cirq/circuits/moment.py:680]: Is there a generic way to get a seq of ops from a thing? It'd be nice to be able to check commutativity between various things that can hold operations, rather than limit to moments or single ops here.

File: cirq-core/cirq/ops/raw_types.py [hunk 4]
```diff
@@ -688,41 +688,7 @@ def _commutes_(
...
-        if np.prod(shape, dtype=np.int64) > 2**10:
-            return NotImplemented  # pragma: no cover
-
-        m12 = protocols.unitary_protocol.unitary(circuit12, default=None)
-        m21 = protocols.unitary_protocol.unitary(circuit21, default=None)
-        if m12 is None or m21 is None:
-            return NotImplemented
-
-        return np.allclose(m12, m21, atol=atol)
+        return _operations_commutes_impl([self], [other], atol=atol)
```
Comment (daxfohl) [inline cirq-core/cirq/ops/raw_types.py:691]: Should this also check commutes(other, self) for consistency?

File: cirq-core/cirq/ops/raw_types.py (hunk 2 above)
Comment (pavoljuhas) [reply cirq-core/cirq/ops/raw_types.py:1117]: Ack.  `protocols.control_keys` calls on the `_control_keys_` method, which is available in [ClassicallyControlledOperation](https://github.com/quantumlib/Cirq/blob/a55b67516d6033028d360ca6c16c8bd1a822aa9e/cirq-core/cirq/ops/classically_controlled_operation.py#L239-L243), but I don't see a public method there or in the base class that would return the same.

On another page, it is faster to build the set of ops1_qubits with `set.union()` compared to a double-loop iteration so I am going to change that.

File: cirq-core/cirq/circuits/moment.py (hunk 3 above)
Comment (pavoljuhas) [reply cirq-core/cirq/circuits/moment.py:680]: Ack.  The `_operations_commutes_impl` assumes that the compared lists of operations have no double-acted qubits which is guaranteed for Moment.operations.  We already fix and extend the previous API here.  I'd rather leave commutation with other types for later when we run into such needs.

File: cirq-core/cirq/ops/raw_types.py (hunk 4 above)
Comment (pavoljuhas) [reply cirq-core/cirq/ops/raw_types.py:691]: AFAICT, the `_operations_commutes_impl` is symmetric in its arguments.
The `protocol.commutes` will also do it for us [anyway](https://github.com/chamodtharakaperera/Cirq/blob/0551f85c5beb26e17187f6f51e3cfd6536ac1327/cirq-core/cirq/protocols/commutes_protocol.py#L160-L174).

**Review by pavoljuhas — APPROVED** [review]

**Comment by pavoljuhas** [comment]
Thank you @chamodtharakaperera and @daxfohl !

File: cirq-core/cirq/circuits/moment.py (hunk 3 above)
Comment (daxfohl) [reply cirq-core/cirq/circuits/moment.py:680]: It occurred to me that maybe `OP_TREE` would be a better input type for the new function, since it supports single ops or op lists. Though it might be a teensy bit slower. Anyway, minor point.
